### PR TITLE
Change Singularity support to use Docker images

### PR DESCRIPTION
**Pull request by kjsanger**
Add installation Makefile for proxy shell scripts to run executables in Docker containers. Initially, only iRODS clients are supported.

Update the main README.

### CODE REVIEW

**Comment by dkj** [comment]
Do we need to worry about where `singularity` will cache the image? Is that not the business of this code/tooling - it should just observe any user's `SINGULARITY_CACHEDIR`? Or, given we're installing wrapper scripts, should we not also get the container image cached? If so, do we use the `SINGULARITY_CACHEDIR` or explicitly tell singularity where to generate and use the sif (that I presume it would otherwise autogenerate)?

Meta question: does this matter for this PR, just punt this to a further refinement?

**Comment by kjsanger** [comment]
> Do we need to worry about where `singularity` will cache the image? Is that not the business of this code/tooling - it should just observe any user's `SINGULARITY_CACHEDIR`? Or, given we're installing wrapper scripts, should we not also get the container image cached? If so, do we use the `SINGULARITY_CACHEDIR` or explicitly tell singularity where to generate and use the sif (that I presume it would otherwise autogenerate)?

Yes, we do need to be aware of where the images will be cached.

No, I don't think it's the job of the wrapper. It's best if it observes standard behaviour (`SINGULARITY_CACHEDIR`). Cache location and capacity will be different for an HPC cluster, an OpenStack VM, a CI runner, a dev laptop.

> Meta question: does this matter for this PR, just punt this to a further refinement?

I don't think it matters because it's not the wrapper's job to have an opinion about it.

File: singularity/Makefile [hunk 1]
```diff
@@ -1,20 +1,29 @@
-URL=$(patsubst %@,%,$(shell git ls-remote --get-url origin))
-TAG=$(shell git describe)
+TAG=?$(shell git describe --always)
```
Comment (dkj) [inline singularity/Makefile:1]:
```suggestion
TAG?=$(shell git describe --always)
```

**Comment by kjsanger** [comment]
I think the same applies to `SINGULARITY_BIND` (for bind points) as for the other `SINGULARITY_*` environment variables. We're likely to want the same settings for all the wrappers in a particular environment, but what those settings should be varies by environment. Best not to have them set explictly in each wrapper script.

The wrappers are not trying to hide the fact that they're using Singularity - I think some documentation would help here (that these variable may need to be set). However, you can use them to make a deployment that _does_ hide the fact that they're using Singularity e.g. by deploying in an Environment Module on a cluster which sets all the relevant `SINGULARITY_*` variables when the module is loaded.

**Review by dkj — APPROVED** [review]
This is useful, both practically and for driving deployment architecture thought.

What it has made me consider:
- does this wrapper functionality belong in this repo: images likely to be built in CI, but this Makefile is run in a deployment of an image from an image repo
- should the singularity image be created at that deployment time and explicitly referenced in the script
  - should the singularity binary similarly be explicitly in the wrappers (so no need to add it onto path first)
  - i.e. make singularity use more transparent and tightly defined for the deployment
- should the wrapper scripts work without the need for their path to be in the `PATH`

**Comment by dkj** [comment]
> Best not to have them set explicitly in each wrapper script.

Agree [for current philosophy of these wrapper scripts].

> The wrappers are not trying to hide the fact that they're using Singularity 

Not so sure about this. However, okay to roll with this as is.

**Comment by kjsanger** [comment]
>     * does this wrapper functionality belong in this repo: images likely to be built in CI, but this Makefile is run in a deployment of an image from an image repo

Not necessarily. It's just here because I needed somewhere to put it and didn't want to fragment the container-related material into more git repos (we have too many repos and there is overhead for maintaining new ones e.g. new CI and ongoing maintenance of that, plus deployment gets more complicated).

The Makefile itself is a straw man - it was just a way to make it easy for people to install the wrappers so that the principle of using them could be tested.

>     * should the singularity image be created at that deployment time and explicitly referenced in the script

Do you mean the Docker image, or are you suggesting we materialise a Singularily image and store that in addition?

>       * should the singularity binary similarly be explicitly in the wrappers (so no need to add it onto path first)

The environment for a particular pipeline deployment should certainly be "frozen" permanently somewhere. I'm not sure putting part of that state (the Singularity binary path) in the wrappers is helpful. I can still change the version of Singularity I'm using by changing the `PATH` to pick up another wrapper also called e.g. `iget` which doesn't pin the Singularity version, or wraps a native `iget`, or anything I like. It's reliant on the environment itself, so doesn't really bring any greater assurance.

>       * i.e. make singularity use more transparent and tightly defined for the deployment

Using a container can also be entirely transparent by setting environment variables. I'm not sure it makes it more tightly defined, either, for the reason above.

>     * should the wrapper scripts work without the need for their path to be in the `PATH`

Can you expand on why? Are you suggesting all the wrappers be called using absolute paths?

**Comment by dkj** [comment]
| | should the singularity image be created at that deployment time and explicitly referenced in the script
| Do you mean the Docker image, or are you suggesting we materialise a Singularily image and store that in addition?

Whatever the container running tool would otherwise use when caching e.g. with singularity it will dynamically create a sif file. I'd suggest caching that sif file "next to" the place the wrapper scripts are deployed.

**Comment by dkj** [comment]
|  It's reliant on the environment itself, so doesn't really bring any greater assurance.
I was valuing not having the dependency on the environment (`PATH`) for `singularity`

**Comment by dkj** [comment]
| Can you expand on why? Are you suggesting all the wrappers be called using absolute paths?

I value the ability of the command to be able to be called with an absolute path rather than have to change `PATH` for it to work. It can avoid the need for another wrapper script rather than symlink if you want to construct a (snapshot-like) directory (rather than relying on env setup procedures) with the tools required for jobs job.
```
dj3@seqfarm3-head1:~/repos$ file /software/npg/20230221/bin/gatk
/software/npg/20230221/bin/gatk: POSIX shell script, ASCII text executable
dj3@seqfarm3-head1:~/repos$ cat /software/npg/20230221/bin/gatk
#!/bin/sh
export PATH=/software/sciops/pkgg/gatk/4.1.3.0/bin/:$PATH
exec /software/sciops/pkgg/gatk/4.1.3.0/bin/gatk "$@"
dj3@seqfarm3-head1:~/repos$ file /software/npg/20230221/bin/baton-do 
/software/npg/20230221/bin/baton-do: symbolic link to /software/sciops/pkgg/baton/4.0.0/bin/baton-do
```

**Comment by kjsanger** [comment]
> I value the ability of the command to be able to be called with an absolute path rather than have to change PATH for it to work.

If the wrappers need to be called from arbirtrary environments where `singularity` may not be on the `PATH`, I can add a build/install parameter so that the absolute `singularity` path is set on wrapper install (e.g. written into the wrapper by processing a template).

Likewise, if there are other runtime parameters that need to be set (e.g. image cache location, default bind points) then they can written into the wrapper via a build parameter.

If those parameters are not supplied (to `make` or whatever's installing the wrapper) then you get the default (i.e. the Singularity executable is`singularity` and the cache and bind points are defaults)

**Pull request merged**